File: AdventOfCode/1/1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

// Function to read numbers from a file into a vector
void readFile(const string &filename, vector<double> &numbers)
{
    ifstream file(filename);
    double num;
    while (file >> num)
    {
        numbers.push_back(num);
    }
    file.close();
}

// Function to calculate the relative difference between two numbers
double relativeDifference(double a, double b)
{
    return abs(a - b) / max(abs(a), abs(b));
}

int main()
{
    // Define file names
    string file1 = "/Users/aditya/Documents/CP/AdventOfCode/1/output_1.txt";
    string file2 = "/Users/aditya/Documents/CP/AdventOfCode/1/output_2.txt";

    // Vectors to store numbers from the files
    vector<double> numbers1, numbers2;

    // Read numbers from both files
    readFile(file1, numbers1);
    readFile(file2, numbers2);

    // Check if both files contain the same number of elements
    if (numbers1.size() != numbers2.size())
    {
        cerr << "Error: The files do not have the same number of elements!" << endl;
        return 1;
    }

    // Sort both vectors
    sort(numbers1.begin(), numbers1.end());
    sort(numbers2.begin(), numbers2.end());

    // Variable to store the sum of relative differences
    int totalRelativeDifference = 0;

    // Calculate the relative differences and sum them up
    for (size_t i = 0; i < numbers1.size(); ++i)
    {
        totalRelativeDifference += abs(numbers1[i] - numbers2[i]);
    }

    // Output the sum of relative differences
    cout << "Total Sum of Relative Differences: " << totalRelativeDifference << endl;

    return 0;
}

File: AdventOfCode/1/1.py
# Define the file paths
input_file = '/Users/aditya/Documents/CP/AdventOfCode/1/input.txt'  # Input file containing the numbers
output_file_1 = '/Users/aditya/Documents/CP/AdventOfCode/1/output_1.txt'  # Output file for the first column
output_file_2 = '/Users/aditya/Documents/CP/AdventOfCode/1/output_2.txt'  # Output file for the second column

# Read the input file and separate the numbers
with open(input_file, 'r') as file:
    lines = file.readlines()

# Initialize lists for the two columns
column_1 = []
column_2 = []

# Process each line to separate the columns
for line in lines:
    # Split by spaces or tabs and strip any extra whitespace
    numbers = line.strip().split()
    if len(numbers) == 2:
        column_1.append(numbers[0])
        column_2.append(numbers[1])

# Write the first column to the output file
with open(output_file_1, 'w') as file1:
    for number in column_1:
        file1.write(number + '\n')

# Write the second column to the output file
with open(output_file_2, 'w') as file2:
    for number in column_2:
        file2.write(number + '\n')

print("Results have been written to", output_file_1, "and", output_file_2)
